07_leukemia_patient/workflow/workflow.sh

```diff
@@ -0,0 +1,232 @@
+       --left_fq $WD/sorted_duplicate_r2.fastq.gz \
+       --genome_lib_dir $STAR_FUSION_GENOME \
+       --output_dir $WD/star_fusion/output > log.txt
+
+echo "STAR_fusion finished"
+
+# adding the CB and UB tag to the bwa mem2 bam file and generating count table
+
+cd $WD/bwa_mem2/output
+
+samtools view -b -F 4 bwa_mem2.sorted.bam > mapped.bam
+
+# removed unmapped reads
+
+samtools view -H mapped.bam > header.txt
+samtools view mapped.bam | awk 'BEGIN {OFS="\t"} {split($1, a, /[;,]/); print a[1], $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, a[2], a[3]}' | cat header.txt - | samtools view -Sbh > annotated_bwa_mem2.sorted.bam
+
+rm header.txt mapped.bam
+
+# UMI deduplication based on CB and UMI and gene name field
+
+echo "# non deduplicated reads bwa mem2"
+
+samtools view annotated_bwa_mem2.sorted.bam | cut -f 1 | wc -l
+
+# bwa mem2 does not require deduplication as the reads are already primary alignments but running anyway
+
+samtools view annotated_bwa_mem2.sorted.bam | awk 'BEGIN{OFS="\t"}{print $1,$3,$16,$17}' | sort -k3,4 | uniq | cut -f1 > dedup_read_names.txt
+samtools view -H annotated_bwa_mem2.sorted.bam > header.txt
+samtools view annotated_bwa_mem2.sorted.bam | grep -Ff dedup_read_names.txt | cat header.txt - | samtools view -Sbh > deduplicated_annotated_bwa_mem2.sorted.bam
+
+echo "# deduplicated reads bwa mem2"
+wc -l dedup_read_names.txt
+
+#rm dedup_read_names.txt header.txt annotated_bwa_mem2.sorted.bam
+
+# count table for bwa_mem2 --> based on counting the occurences for each gene, cell and count. The .fastqs are already deduplicated so don't need to repeat. 
+# column 1: gene id, column 2: barcode id, column 3: counts
+# only count the primary alignment, even if multimapped --> one alignment for mapped read
+
+mkdir -p $WD/count_tables/bwamem2
+cd $WD/count_tables/bwamem2
+
+echo -e "Counts\tGene_id\tBarcode" > counts_bwa_mem2.txt
+
+samtools view $WD/bwa_mem2/output/deduplicated_annotated_bwa_mem2.sorted.bam | cut -f3,16 | sort | uniq -c >> counts_bwa_mem2.txt
+
+# do the same for STAR fusion based on the Chimeric.out.junction
+# only count the ones on the + strand
+# here we are reporting multiple alignments for the same read --> depending if for the same read multiple possible fusions were detected
+# we are also reporting things other than BCR:ABL if we don't filter for the chromosomes --> only take chr22 and chr9
+
+cd $WD/star_fusion/output
+
+# don't want the truncated reads
+
+# just want the ones from chr9 and chr22 
+
+bcr_start=23179704
+bcr_end=23318037
+abl_start=130713016
+abl_end=130887675
+
+awk '$2 >23179704  || $2 <23318037 || $4 > 130713016 || $4 < 130887675' Chimeric.out.junction > sub_Chimeric.out.junction
+
+# UMI deduplication based on CB UB and start / end position
+
+less sub_Chimeric.out.junction | grep -v "-" | grep "chr9" | grep "chr22" | awk 'BEGIN {OFS="\t"} {split($10, a, /[;,]/); print a[1],$1""$4"_"$1"_"$2"__"$4"_"$5,a[2],a[3]}' > annotated_sub_Chimeric.out.junction
```

Comment:
adding apaginato (less) is a bad practice, just grep the file as first step. Applies to all chained pipes starting with a 'less' or 'cat'.